tests/vmemcache_test_mt.c

```diff
@@ -0,0 +1,271 @@
+	}
+
+	run_threads(n_threads, threads, ctx);
+
+	printf("%s: PASSED\n", __func__);
+}
+
+/*
+ * on_evict_cb -- (internal) 'on evict' callback for run_test_get
+ */
+static void
+on_evict_cb(VMEMcache *cache, const char *key, size_t key_size, void *arg)
+{
+	int *cache_is_full = arg;
+
+	*cache_is_full = 1;
+
+	printf("on_evict_cb: evicted the key: %llu\n",
+		*((unsigned long long *)key));
+}
+
+/*
+ * run_test_get -- (internal) run test for vmemcache_get()
+ */
+static void
+run_test_get(VMEMcache *cache, unsigned n_threads, os_thread_t *threads,
+		struct context *ctx)
+{
+	int cache_is_full = 0;
+
+	vmemcache_callback_on_evict(cache, on_evict_cb, &cache_is_full);
+
+	unsigned long long i = 0;
+	while (!cache_is_full) {
+		if (vmemcache_put(ctx->cache, (char *)&i, sizeof(i),
+					ctx->buffs[i % ctx->nbuffs].buff,
+					ctx->buffs[i % ctx->nbuffs].size))
+			FATAL("ERROR: vmemcache_put: %s", vmemcache_errormsg());
+		i++;
+	}
+
+	unsigned long long ops_count = i;
+
+	vmemcache_callback_on_evict(cache, NULL, NULL);
+
+	for (unsigned i = 0; i < n_threads; ++i) {
+		ctx[i].thread_routine = worker_thread_get;
+		ctx[i].ops_count = ops_count;
+	}
+
+	run_threads(n_threads, threads, ctx);
+
+	printf("%s: PASSED\n", __func__);
+}
+
+
+int
+main(int argc, char *argv[])
+{
+	int ret = -1;
+
+	if (argc < 2) {
+		fprintf(stderr, "usage: %s dir-name\n", argv[0]);
+		exit(-1);
+	}
+
+	const char *dir = argv[1];
+
+	/* default values of parameters */
+	unsigned n_threads = 10;
+	unsigned ops_count = 100000;
+	unsigned nbuffs = 10;
+	size_t min_size = 8;
+	size_t max_size = 64;
+
+	VMEMcache *cache = vmemcache_new(dir, VMEMCACHE_MIN_POOL,
+				VMEMCACHE_MIN_FRAG, VMEMCACHE_REPLACEMENT_LRU);
+	if (cache == NULL)
+		FATAL("vmemcache_new: %s (%s)", vmemcache_errormsg(), dir);
+
+	struct buffers *buffs = calloc(nbuffs, sizeof(*buffs));
+	if (buffs == NULL) {
+		FATAL("out of memory");
+		goto exit_delete;
+	}
+
+	srand((unsigned)time(NULL));
+
+	for (unsigned i = 0; i < nbuffs; ++i) {
+		/* generate N random sizes (between A – B bytes) */
+		buffs[i].size = min_size +
+				(size_t)rand() % (max_size - min_size + 1);
+
+		/* allocate a buffer and fill it for every generated size */
+		buffs[i].buff = malloc(buffs[i].size);
+		if (buffs[i].buff == NULL) {
+			FATAL("out of memory");
+			goto exit_free_buffs;
+		}
+
+		memset(buffs[i].buff, 0xCC, buffs[i].size);
+	}
+
+	os_thread_t *threads = calloc(n_threads, sizeof(*threads));
+	if (threads == NULL) {
+		FATAL("out of memory");
+		goto exit_free_ctx;
+	}
+
+	struct context *ctx = calloc(n_threads, sizeof(*ctx));
+	if (ctx == NULL) {
+		FATAL("out of memory");
+		goto exit_free_buffs;
+	}
+
+	for (unsigned i = 0; i < n_threads; ++i) {
+		ctx[i].cache = cache;
+		ctx[i].buffs = buffs;
+		ctx[i].nbuffs = nbuffs;
+		ctx[i].ops_count = ops_count / n_threads;
+		ctx[i].thread_number = i;
+	}
+
+	/* run all tests */
+	run_test_put(n_threads, threads, ctx);
+	run_test_get(cache, n_threads, threads, ctx);
+
```

Comment:
it would be good to have also a mix workload (get/put) and mix with evict workload.

Reply:
It will be added in the next PR, because it requires this fix: #27 to be merged first.